pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -276,10 +307,193 @@ public String segmentUpload(@QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE
     return response;
   }
 
+  @POST
+  @Path(SegmentCompletionProtocol.MSG_TYPE_COMMIT_END_METADATA)
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
+  public String segmentCommitEndWithMetadata(@QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE_ID) String instanceId,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_SEGMENT_NAME) String segmentName,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_SEGMENT_LOCATION) String segmentLocation,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_OFFSET) long offset,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_MEMORY_USED_BYTES) long memoryUsedBytes,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_BUILD_TIME_MILLIS) long buildTimeMillis,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_WAIT_TIME_MILLIS) long waitTimeMillis,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_ROW_COUNT) int numRows,
+                                             @QueryParam(SegmentCompletionProtocol.PARAM_SEGMENT_SIZE_BYTES) long segmentSizeBytes,
+                                             FormDataMultiPart metadataFiles) {
+    if (instanceId == null || segmentName == null || offset == -1 || segmentLocation == null || metadataFiles == null) {
+      LOGGER.error("Invalid call: offset={}, segmentName={}, instanceId={}, segmentLocation={}", offset, segmentName,
+              instanceId, segmentLocation);
+      // TODO: memoryUsedInBytes = 0 if not present in params. Add validation when we start using it
+      return SegmentCompletionProtocol.RESP_FAILED.toJsonString();
+    }
+
+    SegmentCompletionProtocol.Request.Params requestParams = new SegmentCompletionProtocol.Request.Params();
+    requestParams.withInstanceId(instanceId).withSegmentName(segmentName).withOffset(offset)
+            .withSegmentLocation(segmentLocation).withSegmentSizeBytes(segmentSizeBytes)
+            .withBuildTimeMillis(buildTimeMillis).withWaitTimeMillis(waitTimeMillis).withNumRows(numRows)
+            .withMemoryUsedBytes(memoryUsedBytes);
+    LOGGER.info("Processing segmentCommitEnd:{}", requestParams.toString());
+
+
+    final boolean isSuccess = true;
+    final boolean isSplitCommit = true;
+    SegmentMetadataImpl segmentMetadata = extractMetadataFromInput(metadataFiles, segmentName);
+    // If it fails to extract metadata from the input form, try to download the segment and extract it from the segment.
+    if (segmentMetadata == null) {
+      LOGGER.info("Failed to extract segment metadata for {} from input form, fallback to use the segment file.",
+              segmentName);
+      try {
+        segmentMetadata = extractMetadataFromSegmentFile(segmentName, new URI(segmentLocation));
```

Comment:
we should fail the commit if we cannot extract metadata. If a server posted to this URI it better include metadata in the request.

Reply:
Agree. I suppose the code is already doing what you suggested? In the lines follows:

    // Return failure to server if both extraction efforts fail.
    if (segmentMetadata == null) {
      LOGGER.warn("Segment metadata extraction failure for segment {}", segmentName);
      return SegmentCompletionProtocol.RESP_FAILED.toJsonString();
    }

Reply:
I got what you were saying. Simplify the logic and return failure if there is no files in the input.